### PR TITLE
`Development`: Lazily load notifications when opening sidebar

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

Improve the performance by only loading the amount of unread messages and lazily load notifications once the sidebar is opened.

### Description

This PR adds a new Rest-Endpoint that returns the amount of unread notifications. This is used to display the user if they have unread messages. The actual messages are only loaded once the notification sidebar is opened.
Additionally the lastNotificationRead is now only updated if it actually affects any notification. I.e. if there are no unread messages, the date is not updated via a rest call.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students

1. Log in to Artemis
2. Do something to receive a notification. E.g. wait until a new exercise is released.
3. Check that the bell icon shows unread messages when freshly loading the page.
4. Check that you can then see the notification in the side bar.
5. Check that the notification is marked as read and you no longer have unread notifications even when reloading.

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [notification-sidebar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification-sidebar/notification-sidebar.component.ts.html) | 87.66% | ✅ |
| [notification.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Client-Tests/app/shared/notification/notification.service.ts.html) | 84.37% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [NotificationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4992-3/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/NotificationResource.html) | 93% | ✅ |

### CODE REVIEW

**Review by SolizerCodes — APPROVED** [review]
Works on ts1 👍

**Review by MichaelOwenDyer — DISMISSED** [review]
Looks good to me 👍🏽

**Review by asliayk — APPROVED** [review]
works well on ts2

**Review by SabrinaGlatz — DISMISSED** [review]
tested on ts1, works as expected

**Review by YeetTheFirst21 — APPROVED** [review]
Tested on TS1,

all seems to be working as expected

File: src/main/webapp/app/shared/notification/notification-sidebar/notification-sidebar.component.ts [hunk 1]
```diff
@@ -207,7 +213,13 @@ export class NotificationSidebarComponent implements OnInit {
         return [];
     }
 
-    private loadNotifications(): void {
```
Comment (tobias-lippert) [inline src/main/webapp/app/shared/notification/notification-sidebar/notification-sidebar.component.ts:210]: why is this method no longer private?

File: src/test/javascript/spec/component/shared/notification/notification-sidebar.component.spec.ts [hunk 2]
```diff
@@ -267,8 +270,21 @@ describe('Notification Sidebar Component', () => {
...
         }));
 
+        it('should not update users last notification read when no new notification was read', fakeAsync(() => {
+            notificationSidebarComponent.recentNotificationCount = 0;
+
+            jest.spyOn(notificationSidebarComponent, 'updateLastNotificationRead');
+            jest.spyOn(userService, 'updateLastNotificationRead');
+            const bell = notificationSidebarComponentFixture.debugElement.nativeElement.querySelector('.notification-button');
+            bell.click();
+            tick(2000);
```
Comment (tobias-lippert) [inline src/test/javascript/spec/component/shared/notification/notification-sidebar.component.spec.ts:280]: Do we need the 2000 in the tick?

**Review by tobias-lippert — COMMENTED** [review]
Code looks good, just two questions

File: src/test/javascript/spec/component/shared/notification/notification-sidebar.component.spec.ts (hunk 2 above)
Comment (JohannesStoehr) [reply src/test/javascript/spec/component/shared/notification/notification-sidebar.component.spec.ts:280]: I adapted this from the test above.

File: src/main/webapp/app/shared/notification/notification-sidebar/notification-sidebar.component.ts (hunk 1 above)
Comment (JohannesStoehr) [reply src/main/webapp/app/shared/notification/notification-sidebar/notification-sidebar.component.ts:210]: So that it can be called in the tests

**Review by MaximilianAnzinger — APPROVED** [review]
Code LGTM

File: src/test/java/de/tum/in/www1/artemis/notification/NotificationResourceIntegrationTest.java [hunk 3]
```diff
@@ -5,9 +5,7 @@
 import java.time.ZonedDateTime;
 import java.util.List;
 
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
```
Comment (MaximilianAnzinger) [inline src/test/java/de/tum/in/www1/artemis/notification/NotificationResourceIntegrationTest.java:8]: We decided to use wildcards only if there are at least 4 imports.

**Review by MaximilianAnzinger — COMMENTED** [review]
Other than my one comment recent code changes look good to me.

**Review by MaximilianAnzinger — DISMISSED** [review]
Code LGTM 👍

**Review by asliayk — APPROVED** [review]
reapproved

**Review by tobias-lippert — APPROVED** [review]
Reapprove

**Review by MichaelOwenDyer — DISMISSED** [review]
Reapprove

File: src/main/java/de/tum/in/www1/artemis/web/rest/NotificationResource.java [hunk 4]
```diff
@@ -94,4 +96,36 @@ public ResponseEntity<List<Notification>> getAllNotificationsForCurrentUserFilte
...
         return new ResponseEntity<>(page.getContent(), headers, HttpStatus.OK);
     }
+
+    /**
+     * GET notifications/unread-count: Determines the number of unread notifications according to the users' settings and returns it
+     *
+     * @return the amount of unread notifications
+     */
+    @GetMapping("notifications/unread-count")
+    @PreAuthorize("hasRole('USER')")
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/NotificationResource.java:106]:
```suggestion
    @EnforceAtLeastStudent
```

**Review by Strohgelaender — CHANGES_REQUESTED** [review]

**Review by SolizerCodes — APPROVED** [review]
Reapprove

**Review by tobias-lippert — DISMISSED** [review]
Reapprove after fixing merge conflict

**Review by YeetTheFirst21 — APPROVED** [review]
retested on TS1,

all seems to be working as expected.

**Review by MaximilianAnzinger — DISMISSED** [review]
Code LGTM 👍

File: src/test/javascript/spec/component/shared/notification/notification-sidebar.component.spec.ts [hunk 5]
```diff
@@ -243,6 +244,34 @@ describe('Notification Sidebar Component', () => {
...
         });
+
+        it('should load notifications if none are present', () => {
+            jest.spyOn(notificationSidebarComponent, 'loadNotifications');
+            notificationSidebarComponent.page = 0;
+            notificationSidebarComponent.toggleSidebar();
+            expect(notificationSidebarComponent.loadNotifications).toHaveBeenCalledOnce();
+        });
+
+        it('should not load notifications if some are present', () => {
```
Comment (MaximilianAnzinger) [inline src/test/javascript/spec/component/shared/notification/notification-sidebar.component.spec.ts:255]:
```suggestion
        it('should not load notifications repeatedly', () => {
```
Maybe I misunderstand the description of this test. I would appreciate a name that is more descriptive/easier to understand.

**Review by MaximilianAnzinger — DISMISSED** [review]
Left one comment. Other than that Code LGTM 👍

**Review by MaximilianAnzinger — APPROVED** [review]
Code LGTM 👍

**Review by SolizerCodes — APPROVED** [review]
Still works on ts5

**Review by tobias-lippert — DISMISSED** [review]
Reapprove after adding client tests

**Review by YeetTheFirst21 — DISMISSED** [review]
Retested again on Ts1, all seems to be still working as expected after changes.

**Review by krusche — CHANGES_REQUESTED** [review]
The query for unread seems to be quite slow. We should investigate how to make it faster, e.g. by introducing additional indices.

File: src/main/resources/config/liquibase/master.xml [hunk 6]
```diff
@@ -42,6 +42,8 @@
     <include file="classpath:config/liquibase/changelog/20230613000000_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20230601154000_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20230605115500_changelog.xml" relativeToChangelogFile="false"/>
+    <include file="classpath:config/liquibase/changelog/20230701100700_changelog.xml" relativeToChangelogFile="false"/>
+
```
Comment (MaximilianAnzinger) [inline src/main/resources/config/liquibase/master.xml:46]:
```suggestion
```

**Review by MaximilianAnzinger — APPROVED** [review]
You added an empty line that usually is not in the changelog. Other than that Code looks good.

**Review by MaximilianAnzinger — APPROVED** [review]

**Review by YeetTheFirst21 — APPROVED** [review]
Retested on TS1,

all seems to be working as expected

**Comment by krusche** [comment]
we first change how often notifications are loaded, see #7048 and then reconsider this PR

**Comment by JohannesStoehr** [comment]
Solved differently in #7048

**Pull request closed without merging**